Bug/4912 GitGraph routing and colouring for merges and cherry-picks

guypursey opened the pull request:
## :bookmark_tabs: Summary

Fixes issues with colouring and routing of merges in GitGraph.

Resolves #4912 

## :straight_ruler: Design Decisions

- Changed commit variable names `commit1` and `commit2` to `commitA` and `commitB` respectively, to prevent confusion with `seq` property values (which are also numbers and used to ID the commits).
- Refactored `hasOverlappingCommits` function to be easier to read and renamed it `shouldRerouteArrow`:
  - Used `Array.some()` instead of `Array.filter()` (should increase performance with larger graphs, if needed)
  - Swapped out `Object.keys()` for `Object.values()`
  - It takes a new argument: the name of the branch which would receive the arrow's curve and thereby causing issues for overlapping commits.
- Introduced branch colour assignments to each conditional check for routing.
- New e2e tests (many based on reviews by @nirname and @mathbraga)
- Added more sample diagrams to demo folder show how the graphs should look in different situations.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/development.md#3-update-documentation) is used for all new features.
- [x] :bookmark: targeted `develop` branch

guypursey commented:
E2E tests passing locally but just one test is failing remotely (`other/interactive.spec`).

Took the opportunity to reword some commit messages via rebase interactive and force pushed changes to retrigger test run.

guypursey commented:
@nirname Not sure of etiquette re requesting approval but this one is ready for review now 🙂

nirname commented:
@guypursey No problem at all, ask for review as many times as needed.

I saw you have added demo page. That is good, but not enough. We also need to ensure that the following changes won't break this one. Code Coverage complains about that line you've added. Perhaps you need to add another end-2-end test on top of the existing ones. Integration tests basically make a screenshot and compare it when you do the changes. A couple of samples to cover both `p1.y > p2.y` and the opposite `p1.y <= p2.y` would be enough. Please check if this issue is reproduced for vertical orientation (TB, top-bottom) of gitGraph also.

nirname commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -338,27 +338,24 @@ const drawCommits = (svg, commits, modifyGraph) => {
+const hasOverlappingCommits = (commitA, commitB, allCommits) =>
+  Object.entries(allCommits).some(
+    (commitX) =>
+      commitX.branch === commitA.branch && commitX.seq > commitA.seq && commitX.seq < commitB.seq
```
I do not know if it is intended change, or is there any difference which one to use, I would have to debug the code to say that, but previously it was
```
commit.branch === commit2.branch
```
and now it has turned into
```
commit.branch === commitA.branch
```
which basically means
```
commit.branch === commit1.branch // used to be commit2
```

guypursey replied:
Thanks for comments so far -- great feedback.

That change to the choice of commit branch was deliberate -- here is [the commit in which I correct from `commit2` to `commit1`](https://github.com/mermaid-js/mermaid/pull/4927/commits/fe0555f9d7f1fa85728ca3a04a92d453332559e3) from before I renamed the variables. It solves the particular bug I was focused on.

However, I did find there may be another (slightly separate) bug which is live right now but that this PR does not address, but it would be part of the same line of code. So, while I'm checking your other feedback, I'll take a look to see if it makes sense to address it here.

guypursey replied:
I rewrote (and renamed) this function to make it clearer as well as fixing the issues. Hope this makes more sense now.

nirname commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -388,29 +385,29 @@ const findLane = (y1, y2, depth = 0) => {
     offset = 10;
     // Figure out the color of the arrow,arrows going down take the color from the destination branch
-    colorClassNum = branchPos[commit2.branch].index;
+    colorClassNum = branchPos[(p1.y > p2.y ? commitA : commitB).branch].index;
```
We need more integration tests to cover this case

guypursey replied:
This line has now gone and the `colorClassNum` is defined within existing conditions to make sure `TB` cases are accounted for. So I think it's covered by existing integration tests.

nirname reviewed (changes requested):
We also need to check this solution against TB graph

guypursey commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -338,27 +338,24 @@ const drawCommits = (svg, commits, modifyGraph) => {
+const hasOverlappingCommits = (commitA, commitB, allCommits) =>
+  Object.entries(allCommits).some(
+    (commitX) =>
+      commitX.branch === commitA.branch && commitX.seq > commitA.seq && commitX.seq < commitB.seq
```
I think you're right re `Object.entries()`. It should have been `Object.values()`.

Your fix would also work but I don't think we need the key for anything, so grabbing just the values saves us having to make a second look-up (and with any luck, makes the function clearer).

Good catch 👍

guypursey replied:
I've rewritten and renamed this function now so that its purpose should be clearer. Let me know if you spot anything or have any questions or please resolve the conversation if you are happy.

guypursey commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -338,27 +338,24 @@ const drawCommits = (svg, commits, modifyGraph) => {
+const hasOverlappingCommits = (commitA, commitB, allCommits) =>
+  Object.entries(allCommits).some(
+    (commitX) =>
+      commitX.branch === commitA.branch && commitX.seq > commitA.seq && commitX.seq < commitB.seq
```
@mathbraga Thanks for the review of this. It took me a while to understand the original code (which is partly why I rewrote it to try and make it clearer, and marginally more performant).

`hasOverlappingCommits` should return true when two commits that are being connected have other commits between them on the source branch (i.e, the same branch as `commitA` in my rewrite). This being the case, the arrow's path then needs to be diverted so as to avoid running over the other commits and making the graph unclear.

Hope this makes sense but do keep asking questions if you have any, as it may be that I've missed something.

mathbraga replied:
Yes it makes sense. I was unaware of how the cherry-pick feature works exactly, and that is one of the situations where this case would occur.

(the code you provided in the new demos, with the above proposed change in `hasOverlappingCommits`)
![image](https://github.com/mermaid-js/mermaid/assets/31048764/c8b39c9e-1d77-49a4-b385-a82b8f138e2d)

Sorry for the confusion.

guypursey replied:
Could you resolve this conversation if you are happy with the outcome? Thanks!

guypursey commented:
@nirname @mathbraga I have made a few changes since your last reviews and updated the PR description at the top.

It looks like code coverage has increased slightly since from `develop` branch since my recent changes -- I think the integration tests should cover what's needed anyway. When I run Cypress locally, the only changes I see seem to be what I was trying to fix. Let me know if you spot anything else though.

mathbraga commented:
Hey @guypursey, all looks good for my reviews. I think only @nirname can resolve conversations though.

nirname commented:
Your code breaks something. I have faced this rendering issue
<img src="https://github.com/mermaid-js/mermaid/assets/2842432/e01048d9-727f-4d5e-ab75-4ef996096c76" width=500>
While current version works fine
<img src="https://github.com/mermaid-js/mermaid/assets/2842432/014dec42-0664-4ed7-8f4d-14ae196fda2c" width=500>

guypursey commented:
@nirname Good catch. This wasn't covered by the existing tests so I have two e2e tests and a couple of demos.

I removed a redundant line of code and a condition that was causing this.

mathbraga commented:
@guypursey You missed the `gitGraph TB:` on the first line of your second test body (test 35) for the vertical render.

guypursey commented:
> @guypursey You missed the `gitGraph TB:` on the first line of your second test body (test 35) for the vertical render.

Ah, thanks @mathbraga. Well spotted. I've updated that now.

In looking again, I realised I hadn't specified the commit IDs in each new test. This meant the IDs would show up in diff output as they would change between Cypress runs. I've now followed the practice of other graphs and added an ID for each commit.

guypursey commented:
@nirname I've just rebased this PR from develop to ensure coverage is still okay.

I think I've addressed your request now, so ready for review again when you have time :+1:

nirname commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -388,49 +391,59 @@ const findLane = (y1, y2, depth = 0) => {
       if (p1.x < p2.x) {
+        // Destination commit is on branch position left of source commit
+        // so render arrow leftward with colour of destination branch
+        colorClassNum = branchPos[commitB.branch].index;
```
We have to cover this lines by tests

guypursey replied:
Thanks @nirname. I can see [lines 423-424](https://app.codecov.io/gh/mermaid-js/mermaid/pull/4927/blob/packages/mermaid/src/diagrams/git/gitGraphRenderer.js#L423) and [lines 439-440](https://app.codecov.io/gh/mermaid-js/mermaid/pull/4927/blob/packages/mermaid/src/diagrams/git/gitGraphRenderer.js#L439) are not covered so I'll write tests for those. (I thought they would be covered already because the conditions are not new. When I write the new tests, they should increase coverage overall.)

I can see other areas of the file still need coverage but I think these should be a separate PR/issue so I'll leave them for now. Let me know if you think I've missed any though.

nirname replied:
No need to meet all the demands of code analyzers in one approach. Should we need to fix everything we could before applying other incoming improvements the development would stop. Focus on those lines that are the heart of this change. Don't do extra job.

guypursey replied:
@nirname I think I've resolved this issue now (I hope). There's info in the commit messages about reasoning. I've also increased e2e test coverage slightly.

nirname reviewed (changes requested):
@guypursey Sorry for long response. I have discovered that unnecessary routing still appears in your branch, while it does not appear in current release

Expected behaviour (and current release behavior)
<img src="https://github.com/mermaid-js/mermaid/assets/2842432/ce5316af-3a9b-483e-8a06-3c69504b0a4f" width=300>

Current behaviour (in this branch only)
<img src="https://github.com/mermaid-js/mermaid/assets/2842432/94a64efc-7528-49d0-bed5-113df2cdbf46" width=300>

I know it seems to be minor, nevertheless users rely on the layout, and if it changes suddenly their docs may be ruined. We have measure thrice. Try this code to reproduce:

```
gitGraph:
  commit
  branch x
  checkout main
  commit
  checkout x
  commit
  checkout main
  merge x
```

Colors works as expected, by the way. I am not sure why it changes layout. May be this happens because you have changed `shouldRerouteArrow` function (formerly `hasOverlappingCommits`).

Try investigating it further.

nirname commented:
We are almost there. Good to go after fixing some layout problems. Great job.

guypursey commented:
Updated with rebase to bring this PR in line with develop branch.

guypursey commented:
@nirname Not sure why, but codecov is saying coverage has now decreased slightly since I rebased my branch. I'm never sure how accurate this and the problems seems to be in files indirectly affected. I'll see if I can fix it before you review.

guypursey commented:
@nirname Coverage issue seems to be fixed now. (I just fixed a typo in a comment and pushed it here to trigger the e2e check again.)

nirname commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -388,49 +392,69 @@ const findLane = (y1, y2, depth = 0) => {
+  const p1 = commitPos[commitA.id]; // arrowStart
+  const p2 = commitPos[commitB.id]; // arrowEnd
+  const branchToGetCurve =
+    dir === 'TB'
+      ? p1.x < p2.x
+        ? commitB.branch
+        : commitA.branch
+      : p1.y < p2.y
+      ? commitB.branch
+      : commitA.branch;
```
Should not this code be a part of `shouldRerouteArrow` function? It is not used anywhere else. Would not comparison of `seq` do effectively the same as comparison of x / y?

guypursey replied:
Hi, thanks for taking a look. I originally wanted to keep the number of new arguments passed to `shouldRerouteArrow` to a minimum. But you're right: I could pass it `p1` and `p2` and do the calculation for `branchToGetCurve` inside `shouldRerouteArrow`. I can try that out.

I don't think comparing `seq` would be the same -- that would only tell you whether the one commit comes before or after the other on the branch, but here we need to know whether one commit is above or below the other (or in `TB` to the left or right), to know whether the normal branch curve will be a problem or cause a conflict that needs rerouting.

guypursey replied:
I've made that change now although the coverage is decreased again, and I can't work out why. It seems to be arbitrary... :thinking:

guypursey replied:
@nirname I just can't understand why coverage seems to drop randomly...

The number of files and the percentage change seems to be different each time and have no relation to my changes. Sometimes it is up, sometimes it is down.

Any ideas?

![Coverage decrease across indirectly affected files](https://github.com/mermaid-js/mermaid/assets/1826451/ee553adc-f05b-42a6-b241-adfbe2b783cc)

nirname replied:
No worries about coverage. We will have to investigate it further, but not withing this task. The main idea is that your code is covered by tests.

nirname replied:
@guypursey Why would not seq tell us the order of commits? I mean, whether the go left to right, is there any situation when we have 2 commits with the same position? My thoughts that x and y coordinates are simply the reflection of seq...

nirname replied:
I have not debugged it personally, so I can only imply that x and y depend on seq. Also we can get positions from within the function itself, so your initial idea of keeping minimum amount of arguments is still present.

nirname commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -338,26 +338,34 @@ const drawCommits = (svg, commits, modifyGraph) => {
-      allCommits[key].seq < commit2.seq
-    );
+const shouldRerouteArrow = (commitA, commitB, p1, p2, allCommits) => {
+  const commitBIsFurthest = dir === 'TB' ? p1.x < p2.x : p1.y < p2.y;
```
It would be more portable if a function would not have side effects and closures like `dir`, but since it is only for this files and other anonymous functions do the same, that is ok.

guypursey replied:
@nirname The alternative would be to have `commitBIsFurthest` (which relies on `dir`) calculated in `drawArrow` first, and then passed to `shouldRerouteArrow` instead of `p1, p2`. What do you think?

nirname replied:
I see. I don't know which approach is the right one, I would simply do:
```
const shouldRerouteArrow = (commitA, commitB, allCommits) => {
  const p1 = commitPos[commitA.id];
  const p2 = commitPos[commitB.id];
  const commitBIsFurthest = dir === 'TB' ? p1.x < p2.x : p1.y < p2.y;
  ...
```
simply because of arity reduction.

nirname commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -388,49 +396,61 @@ const findLane = (y1, y2, depth = 0) => {
-  const overlappingCommits = hasOverlappingCommits(commit1, commit2, allCommits);
-  // log.debug('drawArrow', p1, p2, overlappingCommits, commit1.id, commit2.id);
+const drawArrow = (svg, commitA, commitB, allCommits) => {
+  const p1 = commitPos[commitA.id]; // arrowStart
+  const p2 = commitPos[commitB.id]; // arrowEnd
+  const arrowNeedsRerouting = shouldRerouteArrow(commitA, commitB, p1, p2, allCommits);
```
I feel like this function signature can be simplified down to
```
shouldRerouteArrow(commitA, commitB, allCommits);
```
I know that it may seem redundant to get p1 and p2 inside the function once again but the access time of commitPos is O(1) and there is no cases when we have to pass positions independently from commits.

What do you think?

nirname replied:
`shouldRerouteArrow` function violates DRY anyway, because it uses `dir` property. In case we consider this function as a method or instance-related function it can use all object attributes without referencing the object directly (this, self ...etc). In case we consider this function as a pure one, we have to pass everything as arguments including `dir`. Right now this is neither of these two. That is why I have suggested to stick to previous signature.

Btw, I am ok with how it looks now. Simply looking for better options.

nirname reviewed (commented):
Code works, rerouting problem has gone. Coloring is correct. I'd like to elaborate more on this to make the code more elegant

nirname reviewed (approved):
Regardless of further refactoring, approval from me. There is no right approach, so to change it or not is up to you. If you feel like one of the solutions is better than the other, than stick to it. Let PR stay open for a short period of time for others to have a look.

sidharthv96 reviewed (approved):
@guypursey PR looks wonderful!

@nirname can you resolve the conversations if you're satisfied with the responses?
